### PR TITLE
Webpack glslify

### PR DESCRIPTION
**Pull request by adnan-wahab**

### CODE REVIEW

File: webgl-shaders/index.js [hunk 1]
```diff
@@ -26,10 +26,9 @@
 
 var glslify = require('glslify');
 
-var shaders = glslify({
-    sourceOnly: true,
-    vertex: './VertexShader.glsl',
-    fragment: './FragmentShader.glsl'
-});
+var shaders = {
+	  vertex: glslify('./VertexShader.glsl'),
```
Comment (michaelobriena) [inline webgl-shaders/index.js]: spacing

**Comment by TheAlphaNerd** [comment]
lgtm :D
This should allow for basic support in other loaders for now

**Comment by michaelobriena** [comment]
@adnan-wahab Are the changes to the shader meant to be included as well?

**Comment by alexanderGugel** [comment]
I don't really see how this addresses the issue. While this now supports Webpack, we still rely on the glslify transform, which means we rely on glslify-loader. E.g. how does this address usage via AMD?

I thought we were moving away from relying on glslify in library code (and instead moving it into a bundler agnostic build step, like in the previous PR).

**Comment by TheAlphaNerd** [comment]
If someone is using another module loading paradigm such as AMD they will still be able to use the built version of famous without a problem.

This will at least fix support with webpack and I believe jspm. It is enough for us to support our community while we figure out the most graceful way to EOL glslify

**Comment by alexanderGugel** [comment]
@TheAlphaNerd I'm not sure I understand how this is going to support AMD without usage of the global build. The current solution would still require a preprocessing step for those people from my understanding.

**Comment by TheAlphaNerd** [comment]
People using amd are going to require a preprocessor either way was my
point.

**Comment by alexanderGugel** [comment]
Hmmmm.... ok. I don't really see why we can't just put the shader in the repo to be honest (other than rather purist reasons), but maybe I'm missing something. Anyways. If that solves the issue...

**Comment by adnan-wahab** [comment]
@michaelobriena 
yes the changes to the shader are required for this to work

**Comment by alexanderGugel** [comment]
@adnan-wahab Is it because glslify renames the uniforms for some reason? I had that issue, but thought I might be missing something. Still seems kind of a glslify bug to me. If that's the case, we should probably fix that in glslify...

**Comment by jd-carroll** [comment]
+1 for putting it directly in the repo and dumping glslify

Not sure if that's even possible, but it would be great to use grunt/gulp + bower for build and dep management.  Right now I find it near impossible to incorporate that type of build because of the glslify.  You are left in a state where some dep's are managed through bower and others are through node :grimacing:.

I could be completely wrong, but I would love to be able to list famous as a bower dep and then use traditional gulp / grunt.

**Comment by michaelobriena** [comment]
@jd-carroll We could still publish to bower with the GLSL code built already.  This would allow people to make changes to our shader code and use those changes without having to download a different tool to compile the shaders for them.

**Comment by jd-carroll** [comment]
I think the bower + grunt/gulp use case is very important.  Without it, I
think adoption rates will be lower than what they otherwise could be.

**Comment by TheAlphaNerd** [comment]
I very much disagree.

We can easily support bower simply by including the standalone in our
tagged version of the platforms

Aside from this the tooling /module loading pattern of the use is a moot
point.

If a develop has a preference, there is nothing stopping them from sticking
famous in their pipeline. Whether it be a script tag or via require (the
standalone will support amd)

If a develop has no preference or does not have experience they can use our
supported work flow with the famous cli and follow along our documentation
to get started.

If you consider that every in production app will have very different
pipeline needs, we cannot ever hope to offer example projects using every
potential tooling system.

What we cam do, and will, is make sure there is nothing in our build that
blocks our developers from using their tool chain of choice.

**Comment by alexanderGugel** [comment]
@TheAlphaNerd I agree from a conceptual point of view, but I think the truth is that we can easily support a wider range of tool chains by simply checking in the shader code. This seems like a very acceptable cost considering the increase in usability and that our dependency on glslify has been a problem in the past. I think that's a much more flexible solution than forcing people to use the standalone version.

> We can easily support bower simply by including the standalone in our tagged version of the platforms

If we do this, I see no reason why we shouldn't also include the bundled shader code.

**Comment by TheAlphaNerd** [comment]
@alexanderGugel while I agree with your points, I respectfully disagree in practice.

Our code base makes a few assumptions
  * Common JS module loading pattern
  * Browserify build step

This stack allows for us to do some magical things in development, like what we are doing with glslify. It will also support us being able to do something fancy for assertion like you are suggesting in #285

We need to make a higher level decision if we want to rely on AST transformation between development and production. If so, there is no need to check in the GL code. If people want to use our "development" code in production, the only thing stopping them from doing so is their technology stack. If someone chooses to use webpack, then they can either require in our standalone (which will just work), or they can shim the variety of AST manipulations we want to use.

There is no disadvantage to using the standalone build, especially as we can include a version that is unminified and includes source maps. There is nothing stopping them from using the development version either, as they can easily rebuild in-between steps.

The amount of build systems and tooling stacks has no shortage, webpack, JSPM, systemJS, bower, npm, commonJS, AMD, es6.... etc etc etc

We simply cannot support all the edge cases in our development branch, but we can very easily publish a version that is compatible.

If we decide that we need our development branch to be a baseline that works in all systems we are effectively giving up the ability to do AST transformations as part of our development process. Perhaps that is preferred... but I think it is super useful and we should be careful about throwing it away

File: webgl-shaders/index.js (hunk 1 above)
Comment (michaelobriena) [inline webgl-shaders/index.js]: @adnan-wahab can you please fix the spacing?

**Comment by alexanderGugel** [comment]
> Our code base makes a few assumptions
> * Common JS module loading pattern
> * **Browserify build step**

The browserify build step is what I'm worried about. One of the advantages of #285 IMO is that everything also "just works fine" without the transform.

> We simply cannot support all the edge cases in our development branch, but we can very easily publish a version that is compatible.

I completely agree! Relying on glslify by contrast means the user needs to have Node installed when creating custom builds.

I think checking in the generated glslify is really not a big deal, especially if it increases usability in *some* cases. That being said I don't have an extremely strong opinion about that since - as you said - users can always still use the standalone build when in doubt. The only thing that I'm worried about is that we need to special case everything because of glslify (e.g. glslify-loader for Webpack), which is IMO not needed.

**Comment by michaelobriena** [comment]
@alexanderGugel I think our main goal is to support multiple build tools and this does that.  I think moving away from a strict dependency on browserify is a good goal to have but having glslify is great for engine development (which is the point of this repo).

If you don't want our tooling, you can use other tools (because of this change) or just use the standalone as Myles mentioned. I don't think it should be the job of our engineers to maintain personal glsl transform tools just because we want to move off of glslify for no real benefit.

**Comment by michaelobriena** [comment]
landed as e4e0f0f

**Comment by talves** [comment]
As a note regarding gulp and browserify, I have a working [famous-gulp-seed][1] which with initial testing supports `browserify` and the `glslify` dependency.

 - Compatible with the `famous-cli` commands and structure using `public` also.
 - Allows for multiple bundle files (multiple entry points)

What I personally think would be a great add for the `famous-cli` is to allow for the passing of your own seed path on `famous create` by adding a prompt for a seed. The default would be the famous seed.  People with little tooling experience can use a third party seed. Just an idea.

[1]: https://github.com/WhyFamous/famous-gulp-seed

**Comment by trusktr** [comment]
@adnan-wahab Do you have a working webpack example? If so, I'm curious to learn from it. If not, I'll have one soon for everyone. (:

**Pull request closed without merging**